yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+    logDebug(s"Stopping $this")
+    // attempt to stop the refresh thread
+    if (!stopRefreshThread()) {
+      closeQueryClient()
+    }
+
+  }
+
+  /**
+   * Close the query client
+   */
+  def closeQueryClient(): Unit = {
+    // and otherwise, stop the query client
+    logDebug("Stopping Timeline client")
+    timelineQueryClient.close()
+  }
+
+  /**
+   * Is the timeline service (and therefore this provider) enabled.
+   * (override point for tests).
+   *
+   * Important: this is called during construction, so test-time subclasses
+   * will be invoked before their own construction has taken place.
+   * Code appropriately.
+   * @return true if the provider/YARN configuration enables the timeline
+   *         service.
+   */
+  def enabled: Boolean = {
+    _enabled
+  }
+  
+  /**
+   * Get the timeline query client. Used internally to ease testing
+   * @return the client.
+   */
+  def getTimelineQueryClient(): TimelineQueryClient = {
+    timelineQueryClient
+  }
+
+  /**
+   * Set the last exception
+   * @param ex exception seen
+   */
+  private def setLastFailure(ex: Throwable): Unit = {
+    setLastFailure(ex, now())
+  }
+
+  /**
+   * Set the last exception
+   * @param ex exception seen
+   * @param timestamp the timestamp of the failure
+   */
+  private def setLastFailure(ex: Throwable, timestamp: Long): Unit = {
+    this.synchronized {
+      lastFailureCause = Some(ex, new Date(timestamp))
+    }
+  }
+
+  /**
+   * Reset the failure info
+   */
+  private def resetLastFailure(): Unit = {
+    this.synchronized {
+      lastFailureCause = None
+    }
+  }
+
+  /**
+   * Get the last exception
+   * @return the last exception or  null
+   */
+  def getLastFailure(): Option[(Throwable, Date)] = {
+    this.synchronized {
+      lastFailureCause
+    }
+  }
+
+  /**
+   * Query for the connection being healthy
+   * @return
+   */
+  def isHealthy(): Boolean = {
+    healthy.get()
+  }
+
+  /**
+   * Get that the health flag itself. This allows test code to initialize it properly.
+   * Also: if accessed and set to false, it will trigger another health chek.
+   * @return
+   */
+  protected def getHealthFlag(): AtomicBoolean = {
+    healthy;
+  }
+
+  /**
+   * Thread safe accessor to application list
+   * @return
+   */
+  def getApplications(): ApplicationListingResults = {
+    this.synchronized(applications)
+  }
+
+  /**
+   * Thread safe call to update the application results
+   * @param newVal new value
+   */
+  protected def setApplications(newVal: ApplicationListingResults): Unit = {
+    this.synchronized {
+      applications = newVal
+    }
+  }
+
+  /**
+   * Health check to call before any other operation is attempted.
+   * This is atomic, using the `healthy` flag to check.
+   * If the endpoint is considered unhealthy then the healthy flag
+   * is reset to false and an exception thrown.
+   * @return true if the health check took place
+   */
+  protected def maybeCheckHealth(): Boolean = {
```

Comment:
Looking at this method and its use in `listApplications`, I noticed two things:

- The return value of this method is never used.
- `listApplications` will query the ATS regardless of the state of this health check, so I'm not sure I understand what's its purpose?